src/runner.py

```diff
@@ -1,53 +1,158 @@
+Why:
+* Impossible to create runner info that is neither local and remote. This
+is allowed by current design.
+* Type-driven design works better with type checkers. Search for "mypy" in comment for add code
+needed for type checker under current-design.
+"""
+
+
 import logging
 import os
-import subprocess
+import subprocess  # nosec B404
 import time
 import urllib.error
 import urllib.request
 from collections import namedtuple
 from pathlib import Path
 from random import choices
 from string import ascii_lowercase, digits
-
-import requests
-from jinja2 import Environment, FileSystemLoader
+from typing import List, Optional, Sequence, TypedDict
 
 import fastcore.net
 import pylxd
+import pylxd.models
+import requests
 from ghapi.all import GhApi
+from jinja2 import Environment, FileSystemLoader
 from pylxd.exceptions import LXDAPIException, NotFound
 
 logger = logging.getLogger(__name__)
-VMResources = namedtuple("VMResource", ["cpu", "memory", "disk"])
+
+VMResources = namedtuple("VMResources", ["cpu", "memory", "disk"])
+
+
+class RunnerLabel(TypedDict):
+    """Represents label in GitHub API for runner status."""
+
+    id: int
+    name: str
+    type: type
+
+
+class RemoteRunner(TypedDict):
+    """Represents runners in GitHub API."""
+
+    id: int
+    busy: bool
+    status: str
+    labels: List[RunnerLabel]
+
+
+class LxdInstanceConfigSource(TypedDict):
+    """Represents configuration for source image in LXD instance."""
+
+    type: str
+    mode: str
+    server: str
+    protocol: str
+    alias: str
+
+
+class LxdInstanceConfig(TypedDict):
+    """Represents configuration for LXD instance."""
+
+    name: str
+    type: str
+    source: LxdInstanceConfigSource
+    ephemeral: bool
+    profiles: List[str]
 
 
 class RunnerInfo:
-    def __init__(self, name, local, remote):
+    """RunnerInfo stores the info on a runner.
+
+    Attributes:
+        name (str): Name of the runner.
+        local (pylxd.models.Instance): Local LXD instance information of runner.
+        remote (RemoteRunner): GitHub API information of runner.
+    """
+
+    def __init__(
+        self, name: str, local: Optional[pylxd.models.Instance], remote: Optional[RemoteRunner]
+    ) -> None:
+        """Construct a instance of runner information.
+
+        Args:
+            name (str): Name of the runner.
+            local (pylxd.models.Instance): Local LXD instance information of runner.
+            remote (RemoteRunner): GitHub API information of runner.
+        """
         self.name = name
         self.local = local
         self.remote = remote
 
     @property
-    def is_idle(self):
-        return self.is_online and self.remote.busy is False
+    def is_idle(self) -> bool:
+        """Whether the runner is idle.
+
+        Returns:
+            bool: Whether the runner is idle.
+        """
+        # Redundant None check for mypy type check.
+        return self.remote is not None and self.is_online and self.remote["busy"] is False
```

Comment:
Would it be clearer if these were re-ordered to something like `return self.is_online and self.remote is not None and self.remote["busy"] is False`? It just keeps the `remote` checks together